MdePkg/Library/BasePanicLib/PanicLib.c

```diff
@@ -0,0 +1,59 @@
+/** @file
+  Base Panic Library.
+
+  Copyright (c) Microsoft Corporation.
+
+**/
+
+#include <Base.h>
+#include <Library/BaseLib.h>
+#include <Library/PrintLib.h>
+#include <Library/SerialPortLib.h>
+#include <Library/PanicLib.h>
+
+//
+// Define the maximum panic message length that this library supports
+//
+#define MAX_PANIC_MESSAGE_LENGTH  0x100
+
+/**
+  Prints a panic message containing a filename, line number, and description.
+  This is always followed by a dead loop.
+
+  Print a message of the form "PANIC <FileName>(<LineNumber>): <Description>\n"
+  to the debug output device.  Immediately after that CpuDeadLoop() is called.
+
+  If FileName is NULL, then a <FileName> string of "(NULL) Filename" is printed.
+  If Description is NULL, then a <Description> string of "(NULL) Description" is printed.
+
+  @param  FileName     The pointer to the name of the source file that generated the panic condition.
+  @param  LineNumber   The line number in the source file that generated the panic condition
+  @param  Description  The pointer to the description of the panic condition.
+
+**/
+VOID
+EFIAPI
+PanicReport (
+  IN CONST CHAR8  *FileName,
+  IN UINTN        LineNumber,
+  IN CONST CHAR8  *Description
+  )
+{
+  CHAR8  Buffer[MAX_PANIC_MESSAGE_LENGTH];
+
+  //
+  // Generate the PANIC message in ASCII format
+  //
+  AsciiSPrint (Buffer, sizeof (Buffer), "PANIC [%a] %a(%d): %a\n", gEfiCallerBaseName, FileName, LineNumber, Description);
+
+  // Initialize Serial Port to write the panic message
+  SerialPortInitialize ();

```

Comment:
Why is this going to `SerialPortLib`?

A generic interface like `DebugLib` allows flexibility to hook into a number of backends including serial port through an instance like `BaseDebugLibSerialPort` or something also generic like the status code router via `PeiDxeDebugLibReportStatusCode`.

Reply:
I just made it work with SerialPortLib.  To be generic I should rename rename this implementation to BasePanicLibSerialPort. I don't think anything else would need to be changed.

Reply:
Why not use a more generic interface?

Reply:
If you're explicitly calling this a serial port lib instance, it's fine to keep.

Reply:
I understand the generic interface approach but for one function I'm not sure it's worth it over just creating an implementation for each backend.  If you see additional value from the generic interface that I'm missing then I can add it.